proxy/src/control/cache.rs

```diff
@@ -0,0 +1,199 @@
+use std;
+use std::collections::HashSet;
+use std::mem;
+
+/// A cache that supports incremental updates with lazy resetting on
+/// invalidation.
+///
+/// When the cache `c` initially becomes invalid (i.e. it becomes
+/// potentially out of sync with the data source so that incremental updates
+/// would stop working), call `c.reset_on_next_modification()`; the next
+/// incremental update will then replace the entire contents of the cache,
+/// instead of incrementally augmenting it. Until that next modification,
+/// however, the stale contents of the cache will be made available.
+pub struct Cache<T> {
+    values: HashSet<T>,
+    reset_on_next_modification: bool,
+}
+
+pub enum Exists<T> {
+    Unknown, // Unknown if the item exists or not
+    Yes(T),
+    No, // Affirmatively known to not exist.
+}
+
+pub enum CacheChange {
```

Comment:
Once again, this type should derive `Debug` (and `Copy`?) IMHO.

Reply:
I don't care much either way. IMO it's better to add those lazily so that we don't have the compiler spending time generating code that we're not using. We can always make it Copy and/or Debug when we need them to be.